src/interactive-window/editor-integration/codewatcher.ts

```diff
@@ -53,6 +53,10 @@ function getIndex(index: number, length: number): number {
     }
 }
 
+/**
+ * CodeWatchers watch source files, provide code lenses in them, and handle the execution of
+ * the code lenses.
+ */
```

Comment:
The name for this module is `CodeWatcher`, does it listen to `TextDocument` content change and parse the document incrementally, or it scans the whole document each time when the content changes?

Reply:
It's actually more complicated than that. There's an update required event for the code lens factory:
https://github.com/microsoft/vscode-jupyter/blob/6f9fd031aeab67741fc36c1ee2d91a002a80e500/src/interactive-window/editor-integration/codewatcher.ts#L91

That fires when the execution count on the interactive window changes (well because the goto code lens needs to update)

And whenever VS code asks for code lenses.

Reply:
It doesn't actually generate the code lenses itself. That's done by the codelensfactory

Reply:
That's here:
https://github.com/microsoft/vscode-jupyter/blob/cf7da9f9c9d43f621d65c91e46c4a0847ff11cef/src/interactive-window/editor-integration/codeLensFactory.ts#L90